### PR TITLE
Add tls connection options

### PR DESCRIPTION
**Pull request by d-bo**
Signed-off-by: d-bo <vathsven@pm.me>

Upade validOptionNames to also support tls options as of MongoDB 4.2

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [x] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

### CODE REVIEW

**Comment by wesleymilan** [comment]
What about TLS support? I also need this feature, can someone tell me when this pull request will be merged?

**Review by dhmlau — APPROVED** [review]
@d-bo, thanks for the contribution. Your changes LGTM. Would like to get at least one more approval before landing. Thanks.

**Review by agnes512 — APPROVED** [review]
I am okay with the changes. One thing is that `tls` options are available in 4.x, and the connector has 3.x as its dependency.

File: lib/mongodb.js
Comment (achrinza) [inline lib/mongodb.js:252]:
```suggestion
      'tls',
      'tlsCertificateKeyFile',
      'tlsCertificateKeyFilePassword',
      'tlsAllowInvalidCertificates',
      'tlsAllowInvalidHostnames',
      'tlsInsecure',
      'tlsCAFile',
```

**Review by achrinza — CHANGES_REQUESTED** [review]
As what @jannyHou has mentioned, some of these options are for the 4.x versions. Could we specify 3.7-specific options only? I've compiled a table of these options:

| Option | 3.7 | 4.2
| - | - | -
| tls | :heavy_check_mark: | :heavy_check_mark: 
| tlsCertificateFile | :x: | :heavy_check_mark: 
| tlsCertificateKeyFile | :heavy_check_mark: | :heavy_check_mark: 
| tlsCertificateKeyFilePassword | :heavy_check_mark: | :heavy_check_mark: 
| tlsAllowInvalidCertificates | :heavy_check_mark: : | :heavy_check_mark: 
| tlsAllowInvalidHostnames | :heavy_check_mark: : | :heavy_check_mark: 
| tlsInsecure | :heavy_check_mark:  | :heavy_check_mark: 
| tlsCAFile | :heavy_check_mark: | :heavy_check_mark: 
| tlsCertificateSelector | :x: | :x: 

## References
- https://mongodb.github.io/node-mongodb-native/3.7/reference/connecting/connection-settings/ (Slightly incorrect & incomplete)
- https://mongodb.github.io/node-mongodb-native/3.7/api/global.html#MongoClientOptions
- https://mongodb.github.io/node-mongodb-native/4.2/interfaces/MongoClientOptions.html

File: lib/mongodb.js
Comment (achrinza) [inline lib/mongodb.js:234]:
```suggestion
    // See https://github.com/mongodb/node-mongodb-native/blob/3.7/lib/mongo_client.js#L94
```

**Review by achrinza — APPROVED** [review]
Thanks, @d-bo! Just a minor change I forgot to mention.

Could you also squash everything into a single commit?

```
git rebase -i master
```

A suggestion for the final commit message could be `feat: add tls options as of mongo 3.7`.

After that, everything should be ready to merge.

**Comment by dhmlau** [comment]
I just kicked off CI.
It might worth to document the table @achrinza compiled in https://github.com/loopbackio/loopback-connector-mongodb/pull/650#pullrequestreview-816467307. But this is out of the scope of this PR. 

@d-bo, if you'd add that into the README, that would be great. thanks.

**Comment by d-bo** [comment]
@dhmlau could offer this kind of table after Connection properties:

### TLS connection properties

| Property   | Type&nbsp;&nbsp; | Description                                                                                                |
| ---------- | ---------------- | ---------------------------------------------------------------------------------------------------------- |
| tls  | Boolean           | Enable TLS connections                                      |
| tlsCertificateFile   | String           | Specifies the location of a local TLS Certificate                                                                                              |
| tlsCertificateKeyFile       | String           | Specifies the location of a local .pem file that contains either the client's TLS/SSL certificate and key or only the client's TLS/SSL key when tlsCertificateFile is used to provide the certificate.                                                                                         |
| tlsCertificateKeyFilePassword       | String           | The password to decrypt the client private key to be used for TLS connections                                                                          |
| tlsAllowInvalidCertificates   | Boolean           | Specifies whether or not the driver should error when the server’s TLS certificate is invalid                                                                            |
| tlsAllowInvalidHostnames       | Boolean           | Specifies whether or not the driver should error when there is a mismatch between the server’s hostname and the hostname specified by the TLS certificate                                                                                          |
| tlsInsecure        | String           | Disables various certificate validations. |
| tlsCAFile       | String           | A path to file with either a single or bundle of certificate authorities to be considered trusted when making a TLS connection                                                                            |

Should i add something ?

**Comment by dhmlau** [comment]
@d-bo, LGTM. If v3.7 doesn't support `tlsCertificateFile`, it would be helpful to specify that in the table too. Thanks!

**Comment by achrinza** [comment]
@dhmlau It might be better toomit it altogether as it's not usable information for this NPM package. Otherwise, it's presence may add confusion.

**Comment by d-bo** [comment]
### TLS connection properties

| Property   | 3.7 | 4.2 | Type | Description                                                                                                |
| ----------| ----- | --- | ---------------- | ---------------------------------------------------------------------------------------------------------- |
| tls  | &#10004; | &#10004; | Boolean           | Enable TLS connections                                      |
| tlsCertificateFile | &#10005; | &#10004; | String           | Specifies the location of a local TLS Certificate                                                                                              |
| tlsCertificateKeyFile    | &#10004; | &#10004; | String           | Specifies the location of a local .pem file that contains either the client's TLS/SSL certificate and key or only the client's TLS/SSL key when tlsCertificateFile is used to provide the certificate.                                                                                         |
| tlsCertificateKeyFilePassword      | &#10004; | &#10004; | String           | The password to decrypt the client private key to be used for TLS connections                                                                          |
| tlsAllowInvalidCertificates  | &#10004; | &#10004; | Boolean           | Specifies whether or not the driver should error when the server’s TLS certificate is invalid                                                                            |
| tlsAllowInvalidHostnames    | &#10004; | &#10004; | Boolean           | Specifies whether or not the driver should error when there is a mismatch between the server’s hostname and the hostname specified by the TLS certificate                                                                                          |
| tlsInsecure       | &#10004; | &#10004; | String           | Disables various certificate validations. |
| tlsCAFile       | &#10004; | &#10004; | String           | A path to file with either a single or bundle of certificate authorities to be considered trusted when making a TLS connection                                                                            |

@dhmlau  Waiting for signal if it really needs to be updated. By the way, maybe it could be usefull to add just a single TLS connection code example in Creating data source section ? (without TLS connection properties table)
```js
const config = {
  name: 'db',
  connector: 'mongodb',
  url: '',
  host: 'localhost',
  port: 27017,
  user: '',
  password: '',
  database: 'testdb',
  tls: true,
  tlsCertificateKeyFile: '/path/to/pem-file',
  tlsCAFile: '/path/to/ca-file',
};
```

**Comment by achrinza** [comment]
I'm in favour of opening a separate PR to update the docs so that we do not block this PR from landing. Would that be ok with you, @d-bo?

**Comment by d-bo** [comment]
> I'm in favour of opening a separate PR to update the docs so that we do not block this PR from landing. Would that be ok with you, @d-bo?

Sure

**Pull request merged**
